sql/2024/04/image-formats.sql

```diff
@@ -0,0 +1,146 @@
+    UNNEST(technologies) AS technology,
+    UNNEST(technology.info) AS info
+  WHERE
+      technology.technology = cms
+    AND ( version = ""
+    OR ENDS_WITH(version, ".x")
+            AND (STARTS_WITH(info, RTRIM(version, "x"))
+        OR info = RTRIM(version, ".x"))
+    OR info = version ) ) );
+
+CREATE TEMPORARY FUNCTION
+  IS_IMAGE (summary STRING)
+  RETURNS BOOLEAN AS ( LOWER(CAST(JSON_EXTRACT_SCALAR(summary, "$.mimeType") AS STRING)) IN ('image/webp',
+                                                                                             'image/png',
+                                                                                             'image/jpeg',
+                                                                                             'image/avif',
+                                                                                             'image/gif',
+                                                                                             'image/bmp'));
+
+CREATE TEMPORARY FUNCTION
+  GET_LCP_ELEMENT_ATTRIBUTE(custom_metrics STRING,
+                            attribute STRING)
+  RETURNS STRING AS ( (
+  SELECT
+    JSON_EXTRACT_SCALAR(attr, "$.value") AS v
+  FROM
+    UNNEST(JSON_EXTRACT_ARRAY(JSON_EXTRACT(custom_metrics, '$.performance.lcp_elem_stats.attributes'))) AS attr
+  WHERE
+      JSON_EXTRACT_SCALAR(attr, "$.name") = attribute
+  LIMIT
+    1 ) );
+
+CREATE TEMPORARY FUNCTION
+  GET_LCP_IMAGE_ATTRIBUTE_VALUE(custom_metrics STRING,
+                                attribute STRING)
+  RETURNS STRING AS ( (
+  SELECT
+    JSON_EXTRACT_SCALAR(attributes, CONCAT("$.", attribute)) AS v
+  FROM
+    UNNEST(JSON_EXTRACT_ARRAY(custom_metrics, '$.Images')) AS attributes
+  WHERE
+    JSON_EXTRACT_SCALAR(attributes, CONCAT("$.", attribute)) IS NOT NULL
+    AND JSON_EXTRACT_SCALAR(attributes, CONCAT("$.", 'url')) = GET_LCP_ELEMENT_ATTRIBUTE(custom_metrics,
+                                                                                         'url')
+  LIMIT
+    1 ) );
+
+WITH
+  pagesWithLcpImages AS (
+    SELECT
+      date,
+      client,
+      page,
+      GET_LCP_ELEMENT_ATTRIBUTE(custom_metrics,
+                                'url') AS url,
+      GET_LCP_IMAGE_ATTRIBUTE_VALUE(custom_metrics,
+                                    'naturalWidth') AS image_width,
+      GET_LCP_IMAGE_ATTRIBUTE_VALUE(custom_metrics,
+                                    'naturalHeight') AS image_height
+    FROM
+      `httparchive.all.pages`
+    WHERE
+      IS_CMS(technologies,
+             'WordPress',
+             '')
+      AND LOWER(JSON_EXTRACT_SCALAR(custom_metrics, '$.performance.lcp_elem_stats.nodeName')) = 'img'
+      AND GET_LCP_IMAGE_ATTRIBUTE_VALUE(custom_metrics,
+                                        'naturalWidth') IS NOT NULL
+      AND GET_LCP_IMAGE_ATTRIBUTE_VALUE(custom_metrics,
+                                        'naturalHeight') IS NOT NULL
+      AND date = '2024-02-01' ),
```

Comment:
Instead of hard-coding this in two places, it would be great to declare a variable `DATE_TO_QUERY` on top of the file which makes it intuitive to change the date. We haven't been doing this before, but @tunetheweb recommended it in https://github.com/GoogleChromeLabs/wpp-research/pull/112#pullrequestreview-1987913106, and I think it's a great little enhancement.